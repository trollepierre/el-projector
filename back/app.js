var createError = require('http-errors')
var express = require('express')
var bodyParser = require('body-parser')
var path = require('path')
var cookieParser = require('cookie-parser')
var logger = require('morgan')
const cors = require('cors')

var tasksRouter = require('./src/routes/tasks-routes');
var dbRouter = require('./src/routes/db');

var app = express()

app.use(logger('dev'))
app.use(bodyParser.json())
app.use(bodyParser.urlencoded({ extended: false }))
app.use(cookieParser())
app.use(cors())

app.use(express.static(path.join(__dirname, '..', 'front-react', 'build')))

app.use('/tasks', tasksRouter);
app.use('/db', dbRouter);

// catch 404 and forward to error handler
app.use(function (req, res, next) {
  next(createError(404))
})

// error handler
app.use(function (err, req, res, next) {
  // set locals, only providing error in development
  res.locals.message = err.message
  res.locals.error = req.app.get('env') === 'development' ? err : {}

  // render the error page
  res.status(err.status || 500)
})

module.exports = app
