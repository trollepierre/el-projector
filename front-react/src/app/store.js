import { withReducer } from '../store/withReducer';
import { apiService, tokenService } from '../services';

const SET_IS_AUTHENTICATED = 'app/SET_IS_AUTHENTICATED';

const setIsAuthenticated = isAuthenticated => dispatch => {
  if(!isAuthenticated){
    tokenService.removeTokens()
  }
  return dispatch({
    type: SET_IS_AUTHENTICATED, payload: { isAuthenticated }
  });
};

const authenticate = password => async dispatch => {
  try {
    const data = {
      email: 'some@example.org',
      name: password
    };
    const tokens = await apiService.post('login', data)
    await tokenService.saveUserTokens(tokens, data)
    dispatch({ type: SET_IS_AUTHENTICATED, payload: { isAuthenticated: true }})
  } catch (error) {
    dispatch({ type: SET_IS_AUTHENTICATED, payload: { isAuthenticated: false }})
    alert('wrong password')
  }
}

const loginSilently = () => async dispatch => {
  try {
    const refreshToken = tokenService.getRefreshToken();
    if(!refreshToken) return;
    const tokens = await apiService.post('login/token', { refreshToken })
    await tokenService.reauthenticate(tokens);
    await dispatch({ type: SET_IS_AUTHENTICATED, payload: { isAuthenticated: true }})
  } catch (error) {
    await dispatch({ type: SET_IS_AUTHENTICATED, payload: { isAuthenticated: false }})
    await tokenService.removeTokens()
    alert('problem during login silently')
    throw error
  }
}

export const actions = {
  setIsAuthenticated,
  authenticate,
  loginSilently,
};

export const initialState = {
  isAuthenticated: false,
};

export const actionHandlers = {
  [SET_IS_AUTHENTICATED]: (state, action) => {
    const { isAuthenticated } = action.payload;
    return {
      ...state,
      isAuthenticated,
    };
  },
};

export const reducer = withReducer(initialState, actionHandlers);
