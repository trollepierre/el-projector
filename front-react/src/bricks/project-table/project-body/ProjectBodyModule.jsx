import React, { useEffect } from 'react';
import TaskRaw from './task-raw/TaskRaw';
import PropTypes from 'prop-types';

const ProjectBodyModule = ({
    tasks,
    fetchTasks,
  }) => {
  const { data, isLoading, error } = tasks

  useEffect(() => {
    fetchTasks()
  }, [fetchTasks])

  return (
  <tbody className="project-body">
    {isLoading ? (<tr><td>loading</td></tr>) : null}
    {error ? <tr><td>An error has occured</td></tr> : null}
    {(!data) ? (<tr><td>oops not possible?</td></tr>) :
      data.map(task => (<TaskRaw task={task} key={task.id}/>))}
    </tbody>
    );
 };

ProjectBodyModule.displayName = 'ProjectBodyModule';

ProjectBodyModule.propTypes = {
  tasks: PropTypes.exact({
    isLoading: PropTypes.bool,
    error: PropTypes.object,
    data: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number,
        name: PropTypes.string,
        description: PropTypes.string,
        points: PropTypes.number,
        createdDate: PropTypes.string,
        endDate: PropTypes.string,
      })
    ),
  }),
  fetchTasks: PropTypes.func.isRequired,
};

ProjectBodyModule.defaultProps = {
  tasks: {
  isLoading: true,
  error: undefined,
  data: []
  }
};

export default ProjectBodyModule;
