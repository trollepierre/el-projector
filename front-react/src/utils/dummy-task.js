const dummyTask = (task) => ({
  id: 1,
  name: 'Build el-projector',
  description: 'more data about how to do it',
  points: 999, // given by server
  createdDate: '19/05/2019',
  endDate: undefined, // must be done before this date
  optimalDate: undefined, // user would do it at this date
  isDone: false,
  doneDate: undefined, // for log record
  priority: 'HIGHLY_IMPORTANT', // Enum with 5 levels (inherited by default, but can be updated)
  benefit: 'ORGASM_HAPPINESS', // Enum with 5 levels
  neededTime: 'ONE_MONTH', // Enum
  epicId: 'A1', // Yearly happiness epic
  neededMaterial: ['MacTo', 'iMoi', 'MacNo', 'home'],
  subTaskIds: ['TASK10', 'TASK11', 'TASK12'], // checklist of subtasks that depends on it
  masterTaskId: 'TASK0', // this task depends on the master one
  ...task,
})

export { dummyTask }
