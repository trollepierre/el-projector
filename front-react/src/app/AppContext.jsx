import React, { useReducer, createContext, useContext } from 'react';
import PropTypes from 'prop-types';
import { reducer, initialState, actions } from './store';

const AppContext = createContext();

export const useAppContext = () => useContext(AppContext);

const getProps = (state, dispatch) => ({
  ...state,
  setIsAuthenticated: isAuthenticated => actions.setIsAuthenticated(isAuthenticated)(dispatch),
  authenticate: value => actions.authenticate(value)(dispatch),
});

export const AppProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <AppContext.Provider value={{ ...getProps(state, dispatch) }}>
      {children}
    </AppContext.Provider>
  );
};

AppProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default AppContext;
