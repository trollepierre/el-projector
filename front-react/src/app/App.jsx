import React from 'react';
import { Provider } from 'react-redux';
import ProjectPage from '../pages/ProjectPage/ProjectPage';
import { default as createStore } from '../store/createStore';
import { default as reducer } from '../bricks/project-table/project-body/store';
import LoginPage from '../pages/LoginPage';
import { auth } from '../services';

const store = createStore(reducer)

const App = () => {
  if(!auth.isAuthenticated()) return (<LoginPage/>)

  return (
    <div className="App">
      <Provider store={store}>
        <ProjectPage/>
      </Provider>
    </div>
  );
};

export default App;
