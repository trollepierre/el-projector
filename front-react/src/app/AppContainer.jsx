import React, { useEffect } from 'react';
import { AppProvider, useAppContext } from './AppContext';

import LoginPage from '../pages/LoginPage';
import { Header } from '../bricks';
import ProjectPage from '../pages/ProjectPage/ProjectPage';

const AppContainer = () => {
  const { isAuthenticated } = useAppContext();

  if (!isAuthenticated) {
    return (<LoginPage/>);
  }

  return (
    <div className="App">
      <Header/>
      <ProjectPage/>
    </div>
  );
};

const AppModule = () => {
  return (
    <AppProvider>
      <AppContainer/>
    </AppProvider>
  );
};

export default AppModule;
