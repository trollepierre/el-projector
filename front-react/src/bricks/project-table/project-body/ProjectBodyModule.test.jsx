import React from 'react';
import { shallow } from 'enzyme';
import ProjectBodyModule from './ProjectBodyModule';
import { dummyTask } from '../../../utils/dummy-task';
import * as AppContext from '../../../app/AppContext';

describe('ProjectBodyModule', () => {

  beforeEach(() => {
    jest
      .spyOn(AppContext, 'useAppContext')
      .mockImplementation(() => ({ loginSilently: jest.fn() }));

  });
  describe('template', () => {
    it('should match default snapshot', () => {
      // When
      const wrapper = shallow(<ProjectBodyModule fetchTasks={jest.fn()}/>);

      // Then
      expect(wrapper).toMatchSnapshot();
    });

    it('should match snapshot with props', () => {
      // Given
      const props = {
        fetchTasks: jest.fn(),
        isLoading: false,
        error: { message: 'error' },
        data: [dummyTask(), dummyTask()]
      };

      // When
      const wrapper = shallow(<ProjectBodyModule {...props} />);

      // Then
      expect(wrapper).toMatchSnapshot();
    });
  });
});
