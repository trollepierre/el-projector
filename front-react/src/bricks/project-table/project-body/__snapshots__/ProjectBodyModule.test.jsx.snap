// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectBodyModule template should match default snapshot 1`] = `
<tbody
  className="project-body"
>
  <tr>
    <td>
      loading
    </td>
  </tr>
</tbody>
`;

exports[`ProjectBodyModule template should match snapshot with props 1`] = `
<tbody
  className="project-body"
>
  <tr>
    <td>
      loading
    </td>
  </tr>
</tbody>
`;
